### ISSUE TITLE
Polymorphic function value not allowed depending on placement of braces

### ISSUE THREAD

**Issue by jrudolph** [lampepfl/dotty#12323]
Seems to be a variation of #6725:

```scala
type X[T] = [U] => U => U

// These work:
val x1: X[Int] = [U] => (u: U) => u
val x2: X[Int] = [U] => ((u: U) => u)
val x3: X[Int] = [U] => (u: U) => { u }

// This one complains with "Implementation restriction: polymorphic function literals must have a value parameter"
val xe: X[Int] = [U] => { (u: U) => u }
```

**Comment by odersky** [comment]
We should get rid of that implementation restriction.

**Comment by jrudolph** [comment]
> We should get rid of that implementation restriction.

It's not quite clear to me what the actual implementation restriction currently is? Is it that polymorphic function literals are not allowed to be nullary (aside from the type parameters)?

I opened a PR just to fix that syntax anomaly here: https://github.com/lampepfl/dotty/pull/12331. Maybe we should create a separate ticket for the functionality change?

**Comment by bishabosha** [comment]
> I opened a PR just to fix that syntax anomaly here: #12331. Maybe we should create a separate ticket for the functionality change?

I opened https://github.com/lampepfl/dotty-feature-requests/issues/206
